veracruz-mcu-client/vc.c

```diff
@@ -0,0 +1,577 @@
+        mbedtls_pk_free(&vc->client_key);
+        mbedtls_x509_crt_free(&vc->client_cert);
+        free(vc->recv_buf);
+        free(vc->send_buf);
+        return err;
+    }
+
+    mbedtls_debug_set_threshold(4);
+    mbedtls_ssl_conf_dbg(&vc->session_cfg, mbedtls_debug, vc);
+
+    err = mbedtls_ssl_setup(&vc->session, &vc->session_cfg);
+    if (err) {
+        printf("failed to setup SSL session (%d)\n", err);
+        mbedtls_ssl_config_free(&vc->session_cfg);
+        mbedtls_pk_free(&vc->client_key);
+        mbedtls_x509_crt_free(&vc->client_cert);
+        free(vc->recv_buf);
+        free(vc->send_buf);
+        return err;
+    }
+
+    // setup blocking IO functions
+    mbedtls_ssl_set_bio(&vc->session,
+            vc,
+            vc_rawsend,
+            NULL, // no non-blocking recv
+            vc_rawrecv);
+
+    // perform SSL handshake
+    printf("beginning TLS handshake with enclave{%s:%d}\n",
+            VC_SERVER_HOST,
+            VC_SERVER_PORT);
+    printf("policy hash: ");
+    hex(VC_POLICY_HASH, sizeof(VC_POLICY_HASH));
+    printf("\n");
+    printf("client cert hash: ");
+    hex(VC_CLIENT_CERT_HASH, sizeof(VC_CLIENT_CERT_HASH));
+    printf("\n");
+    printf("CA cert hash: ");
+    hex(VC_CA_CERT_HASH, sizeof(VC_CA_CERT_HASH));
+    printf("\n");
+    err = mbedtls_ssl_handshake(&vc->session);
+    if (err) {
+        printf("mbedtls_ssl_handshake failed (%d)\n", err);
+        mbedtls_ssl_free(&vc->session);
+        mbedtls_ssl_config_free(&vc->session_cfg);
+        mbedtls_pk_free(&vc->client_key);
+        mbedtls_x509_crt_free(&vc->client_cert);
+        free(vc->recv_buf);
+        free(vc->send_buf);
+        return err;
+    }
+
+    // success!
+    printf("\033[32mestablished TLS session with enclave{%s:%d}\033[m\n",
+            VC_SERVER_HOST,
+            VC_SERVER_PORT);
+
+    const mbedtls_x509_crt *peer = mbedtls_ssl_get_peer_cert(&vc->session);
+
+    #ifdef VC_DUMP_INFO
+        printf("enclave cert:\n");
+        xxd(peer->raw.p, peer->raw.len);
+    #endif
+
+    // verify runtime hash
+    err = vc_verify_runtime_hash(vc, peer);
+    if (err) {
+        vc_close(vc);
+        return err;
+    }
+
+    k_sleep(Z_TIMEOUT_MS(DELAY*1000));
+    return 0;
+}
+
+int vc_close(vc_t *vc) {
+    mbedtls_ssl_free(&vc->session);
+    mbedtls_ssl_config_free(&vc->session_cfg);
+    mbedtls_pk_free(&vc->client_key);
+    mbedtls_x509_crt_free(&vc->client_cert);
+    free(vc->recv_buf);
+    free(vc->send_buf);
+    return 0;
+}
+
+// helper for encoding dynamic-length bytes/strings
+struct bytes {
+    const void *buf;
+    size_t len;
+};
+
+static bool vc_encode_bytes(
+        pb_ostream_t *stream,
+        const pb_field_iter_t *field,
+        void *const *arg) {
+    if (!pb_encode_tag_for_field(stream, field))
+        return false;
+
+    struct bytes *b = *arg;
+    return pb_encode_string(stream, b->buf, b->len);
+}
+
+int vc_send_data(vc_t *vc,
```

Comment:
Could we get some documentation on these functions? Does this function just handle sending Veracruz `data`, or is it also used to send Veracruz `program`? How would a client retrieve a computation result?

Reply:
These are documented in vc.h, though I can add more. The get_results/send_program functions are currently missing, but easy to add